src/contents/private-smart-contracts-with-solidity-and-circom.mdx

```diff
@@ -0,0 +1,370 @@
+
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.7.0 <0.9.0;
+
+interface ICircomVerifier {
+    function verifyProof(uint[2] calldata _pA, uint[2][2] calldata _pB, uint[2] calldata _pC, uint[4] calldata _pubSignals) external view returns (bool);
+}
+
+contract CircomVoter {
+    ICircomVerifier circomVerifier;
+    uint public publicInput;
+
+    struct Proposal {
+        string description;
+        uint deadline;
+        uint forVotes;
+        uint againstVotes;
+    }
+
+    uint merkleRoot;
+    uint proposalCount;
+    mapping (uint proposalId => Proposal) public proposals;
+    mapping (uint nullifier => bool isNullified) public nullifiers;
+
+    constructor(uint _merkleRoot, address circomVeriferAddress) {
+        merkleRoot = _merkleRoot;
+        circomVerifier = ICircomVerifier(circomVeriferAddress);
+    }
+
+    function propose(string memory description, uint deadline) public {
+        proposals[proposalCount] = Proposal(description, deadline, 0, 0);
+        proposalCount += 1;
+    }
+
+    function castVote(uint[2] calldata _pA, uint[2][2] calldata _pB, uint[2] calldata _pC, uint[4] calldata _pubSignals) public {
+        circomVerifier.verifyProof(_pA, _pB, _pC, _pubSignals);
+        uint nullifier = _pubSignals[0];
+        uint merkleRootPublicInput = _pubSignals[1];
+        uint proposalId = uint(_pubSignals[2]);
+        uint vote = uint(_pubSignals[3]);
+
+        require(block.timestamp < proposals[proposalId].deadline, "Voting period is over");
+        require(merkleRoot == merkleRootPublicInput, "Invalid merke root");
+        require(!nullifiers[nullifier], "Vote already casted");
+
+        nullifiers[nullifier] = true;
+
+        if(vote == 1)
+            proposals[proposalId].forVotes += 1;
+        else if (vote == 2)
+            proposals[proposalId].againstVotes += 1;
+    }
+}
+```
+
+Now, create the first proposal for voting by calling the propose() function. For example, you can test by creating a vote with "Should we eat pizza?" as the description and with 1811799232 as the deadline, which expires in 2027.
+
+Next, let's generate a proof in the format required to verify it in Remix.
+
+```bash
+snarkjs groth16 prove proveVote_0001.zkey witness.wtns proof.json public.json
+snarkjs groth16 verify verification_key.json public.json proof.json
+snarkjs generatecall
+```
+
+Let's pass the result from the terminal as a parameter in Remix, and we'll see how the vote was executed by accessing the data of proposal 0 through the proposals mapping.
+
+<div className="text-center">
+![Votos anónimos con Circom y Solidity](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/vxtljhmmkyv2zh30i7tz.png)
+</div>
+
+We observe that our vote was counted without revealing who the sender was. Try to cast the same vote again, and you'll see that it won't be possible—the transaction will revert. This is because we nullified the vote so that each voter can only cast one vote.
```

Comment:
Maybe end with a wrap up paragraph of some sort on what they learned today and future practice?
The ending here feels a bit abrupt